crates/walrus-service/src/node/blob_event_processor.rs

```diff
@@ -0,0 +1,311 @@
+    }
+
+    /// Processes a blob deleted event.
+    #[tracing::instrument(skip_all)]
+    async fn process_blob_deleted_event(
+        &self,
+        event_handle: EventHandle,
+        event: BlobDeleted,
+    ) -> anyhow::Result<()> {
+        let blob_id = event.blob_id;
+
+        if let Some(blob_info) = self.node.storage.get_blob_info(&blob_id)? {
+            if !blob_info.is_certified(self.node.current_epoch()) {
+                self.node
+                    .blob_retirement_notifier
+                    .notify_blob_retirement(&blob_id);
+                self.blob_sync_handler
+                    .cancel_sync_and_mark_event_complete(&blob_id)
+                    .await?;
+            }
+            // Note that this function is called *after* the blob info has already been updated with
+            // the event. So it can happen that the only registered blob was deleted and the blob is
+            // now no longer registered.
+            // We use the event's epoch for this check (as opposed to the current epoch) as
+            // subsequent certify or delete events may update the `blob_info`; so we cannot remove
+            // it even if it is no longer valid in the *current* epoch
+            if !blob_info.is_registered(event.epoch) {
+                tracing::debug!("deleting data for deleted blob");
+                // TODO (WAL-201): Actually delete blob data.
+            }
+        } else {
+            tracing::warn!(
+                walrus.blob_id = %blob_id,
+                "handling `BlobDeleted` event for an untracked blob"
+            );
+        }
+
+        event_handle.mark_as_complete();
+
+        Ok(())
+    }
+
+    /// Processes a blob invalid event.
+    #[tracing::instrument(skip_all)]
+    async fn process_blob_invalid_event(
+        &self,
+        event_handle: EventHandle,
+        event: InvalidBlobId,
+    ) -> anyhow::Result<()> {
+        self.node
+            .blob_retirement_notifier
+            .notify_blob_retirement(&event.blob_id);
+        self.blob_sync_handler
+            .cancel_sync_and_mark_event_complete(&event.blob_id)
+            .await?;
+        self.node.storage.delete_blob_data(&event.blob_id).await?;
+
+        event_handle.mark_as_complete();
+        Ok(())
+    }
+}
+
+/// Blob event processor that processes blob events. It can be configured to process events
+/// sequentially or in parallel using background workers.
+#[derive(Debug, Clone)]
+pub struct BlobEventProcessor {
+    node: Arc<StorageNodeInner>,
+
+    // Background processors that process events in parallel.
+    background_processor_senders: Vec<UnboundedSender<(EventHandle, BlobEvent)>>,
+    _background_processors: Vec<Arc<JoinHandle<()>>>,
+
+    // When there are no background workers, we use a sequential processor to process events using
+    // this processor. This is to keep the same behavior as before BackgroundEventProcessor.
+    sequential_processor: Option<Arc<BackgroundEventProcessor>>,
```

Comment:
Nit/remark/suggestion (maybe for a follow-up): Given that either one or the other is set, this would be a good match for an enum instead of multiple fields.
OTOH, we will probably remove the sequential option anyway (or use a single background process for it), so not worth investing many efforts in it.